### DISCUSSION TITLE
Serializing a HashMap as a list of tuples while specifiying both types

### DISCUSSION THREAD

**Discussion by Lehona** [jonasbb/serde_with#231] [Q&A]
When trying to serialize a HashMap as a list of tuples (i.e. a `Vec<(_, _)>`) specifying both types causes serde_as to fail. Example:

Fails:
```
    #[serde_as(as = "Vec<(String, LocalPy)>")]
    pub text: HashMap<String, Py<Inner>>,
``` 

Works: 

```
    #[serde_as(as = "Vec<(_, LocalPy)>")]
    pub text: HashMap<String, Py<Inner>>,
``` 

Where `LocalPy` is an adapter for any `Py<T>`.

I just figured out that leaving out one of the types works, but in the (rare) case that one needs to specify an adapter for both key and value, this wouldn't work. I don't see a reason why it wouldn't (I assume "inference" would infer the key type to be String anyway), so it's most likely just an oversight :)

**Comment by jonasbb** [comment]
Hi @Lehona, I haven't figured out how to document this yet. Two transforms will work without problem (e.g., [this test](https://github.com/jonasbb/serde_with/blob/master/tests/serde_as.rs#L295)).
The problem here is that `String` does not implement `SerializeAs`, so the `String` <-> `String` conversion is the problem.
Every time you want to use the normal `Serialize` implementation of a type, you have to write `_` or [`Same`](https://docs.rs/serde_with/1.6.0/serde_with/struct.Same.html). The `_` will simple be converted into `Same`.

A blanket implementation of `SerializeAs` for all types implementing `Serialize` is not possible, due to overlapping implementations. For each type such an implementation is possible, but this would special case the standard library and would probably miss some types, which is why I haven't done it.

There is some more context in #201 and #202.
